File: 3.Longest Substring Without Repeating Characters.2.py
import collections
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        cnt = collections.Counter()
        res = l = 0
        for r, c in enumerate(s):
            cnt[c] += 1
            while cnt[c] == 2:
                cnt[s[l]] -= 1
                l += 1
            res = max(res, r - l + 1)
        return res

File: 3.Longest Substring Without Repeating Characters.2.c
#define MAX(x, y) (((x) > (y)) ? (x) : (y))

int lengthOfLongestSubstring(char * s){
    int cnt[256] = {0};
    int res = 0;
    for (char* r = s, *l = s; *r != '\0'; r++) {
        if (++cnt[*r] == 2) {
            while (cnt[*r] == 2) {
                cnt[*(l++)]--;
            }
        }
        res = MAX(res, r - l + 1);
    }
    return res;
}
